### DISCUSSION TITLE
Cypress Detached from DOM error

### DISCUSSION THREAD

**Discussion by jchi2241** [cypress-io/cypress#28515] [Questions and Help]
We have a flaky test that doesn't seem to be retrying the chain of queries if the final "click" command is called on a detached DOM.
```js
cy.findByRole("button", { name: "Resize" })
   .should(($el) => expect($el).to.not.have.class("disabled"))
   .click()
```
error
```
		cy:command (success):
				assert	expected **<div.components-menu-item.component-menu-item--variant-neutral>** not to have class **disabled**
		cy:command (error):
				click	
...

Errors:
	CypressError: Timed out retrying after 15050ms: `cy.click()` failed because the page updated while this command was executing. Cypress tried to locate elements based on this query:
	
	> <div.components-menu-item.component-menu-item--variant-neutral>
	
	We initially found matching element(s), but while waiting for them to become actionable, they disappeared from the page. Common situations why this happens:
	  - Your JS framework re-rendered asynchronously
	  - Your app code reacted to an event firing and removed the element
	
	You can typically solve this by breaking up a chain. For example, rewrite:
	
	> `cy.get('button').click().click()`
	
	to
	
	> `cy.get('button').as('btn').click()`
	> `cy.get('@btn').click()`
	
	https://on.cypress.io/element-has-detached-from-dom
```

My expectation is that Cypress [retries the queries from the top of the chain](https://docs.cypress.io/guides/core-concepts/retry-ability#Commands-Queries-and-Assertions) (`findByRole`, `should`) if a command (`click`) fails? Does "failure" include "detached DOM" issues? 

Note the following is flake-free:
```js
cy.findByRole("button", { name: "Resize" }).should(($el) => expect($el).to.not.have.class("disabled"));
cy.findByRole("button", { name: "Resize" }).click()
```
Why? 

Relevant versions:
```
"cypress": "12.17.4"
"@testing-library/cypress": "9.0.0"
"@testing-library/dom": "8.6.0"
"@testing-library/react": "12.1.0"
```

We are using Cypress Component tests

**Comment by portyu9** [comment]
Hi @jchi2241 

Your second version is the correct pattern.

This chain:

```js
cy.findByRole("button", { name: "Resize" })
  .should(($el) => expect($el).to.not.have.class("disabled"))
  .click()
```

has an assertion in the middle. Once `.should()` passes, Cypress has successfully resolved and **locked in that subject**.

If React then re-renders the button before `.click()` executes, the element that passed the assertion can now be detached. Cypress cannot go back through that already-passed assertion and obtain the replacement button.

Splitting the chains forces Cypress to perform a fresh query:

```js
cy.findByRole('button', { name: 'Resize' })
  .should('not.have.class', 'disabled')

cy.findByRole('button', { name: 'Resize' })
  .click()
```

If this is an actual disabled button rather than a CSS-only state, I would prefer:

```js
cy.findByRole('button', { name: 'Resize' })
  .should('be.enabled')

cy.findByRole('button', { name: 'Resize' })
  .click()
```

This is not specific to `@testing-library/cypress`. It is Cypress retry semantics combined with the application replacing the DOM node after the assertion passes.

Current Cypress documentation explicitly recommends splitting chains at assertion boundaries when a re-render could make the subject stale.

Please mark this answer as accepted if it helped, thank you.
